torchrl/modules/planners/cem.py

```diff
@@ -0,0 +1,133 @@
+        action_key: str = "action",
+    ):
+        super().__init__(env=env, action_key=action_key)
+        self.planning_horizon = planning_horizon
+        self.optim_steps = optim_steps
+        self.num_candidates = num_candidates
+        self.num_top_k_candidates = num_top_k_candidates
+        self.reward_key = reward_key
+
+    def planning(self, td: TensorDictBase) -> torch.Tensor:
+        batch_size = td.batch_size
+        expanded_original_td = (
+            td.expand(*batch_size, self.num_candidates).view(-1)
+        )
+        flatten_batch_size = batch_size.numel()
+        actions_means = torch.zeros(
+            flatten_batch_size,
+            1,
+            self.planning_horizon,
+            *self.action_spec.shape,
+            device=td.device,
+            dtype=self.env.action_spec.dtype,
+        )
+        actions_stds = torch.ones(
+            flatten_batch_size,
+            1,
+            self.planning_horizon,
+            *self.action_spec.shape,
+            device=td.device,
+            dtype=self.env.action_spec.dtype,
+        )
+        for _ in range(self.optim_steps):
+            actions = actions_means + actions_stds * torch.randn(
+                flatten_batch_size,
+                self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+                device=td.device,
+                dtype=self.env.action_spec.dtype,
+            )
+            actions = actions.view(
+                flatten_batch_size * self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+            )
+            actions = self.env.action_spec.project(actions)
+            optim_td = expanded_original_td.clone()
+            policy = PrecomputedActionsSequentialSetter(actions)
+            optim_td = self.env.rollout(
+                max_steps=self.planning_horizon,
+                policy=policy,
+                auto_reset=False,
+                tensordict=optim_td,
+            )
+            rewards = (
+                optim_td.get(self.reward_key)
+                .sum(dim=1)
+                .reshape(flatten_batch_size, self.num_candidates)
+            )
+            _, top_k = rewards.topk(self.num_top_k_candidates, dim=1)
+            top_k += (
+                torch.arange(0, flatten_batch_size, device=td.device).unsqueeze(
+                    1
+                )
+                * self.num_candidates
+            )
+            best_actions = actions.view(
+                flatten_batch_size,
+                self.num_candidates,
+                self.planning_horizon,
+                *self.action_spec.shape,
+            )[torch.arange(flatten_batch_size), top_k]
+            actions_means = best_actions.mean(dim=1, keepdim=True)
+            actions_stds = best_actions.std(dim=1, keepdim=True)
+        return (actions_means[:, :, 0]).view(*batch_size, *self.action_spec.shape)
```

Comment:
I would use unflatten here, instead of view

Reply:
We actually need view since we are unflattening but also unsqueezing